tpm/src/socket.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright © 2022, Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+
+use crate::TPM_CRB_BUFFER_MAX;
+use anyhow::anyhow;
+use std::io::Read;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::os::unix::net::UnixStream;
+use std::thread::sleep;
+use std::time::{Duration, Instant};
+use thiserror::Error;
+use vmm_sys_util::sock_ctrl_msg::ScmSocket;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Cannot connect to tpm Socket even after retrying for 10 secs")]
+    ConnectToSocket(#[source] anyhow::Error),
+    #[error("Failed to read from socket")]
+    ReadFromSocket(#[source] anyhow::Error),
+    #[error("Failed to write to socket")]
+    WriteToSocket(#[source] anyhow::Error),
+}
+type Result<T> = anyhow::Result<T, Error>;
+
+#[derive(PartialEq)]
+enum SocketDevState {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+pub struct SocketDev {
+    state: SocketDevState,
+    stream: Option<UnixStream>,
+    // Fd sent to swtpm process for Data Channel
+    write_msgfd: RawFd,
+    // Data Channel used by Cloud-Hypervisor
+    data_fd: RawFd,
+    // Control Channel used by Cloud-Hypervisor
+    ctrl_fd: RawFd,
+}
+
+impl Default for SocketDev {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SocketDev {
+    pub fn new() -> Self {
+        Self {
+            state: SocketDevState::Disconnected,
+            stream: None,
+            write_msgfd: -1,
+            ctrl_fd: -1,
+            data_fd: -1,
+        }
+    }
+
+    pub fn init(&mut self, path: String) -> Result<()> {
+        self.connect(&path)?;
+        Ok(())
+    }
+
+    pub fn connect(&mut self, socket_path: &str) -> Result<()> {
+        self.state = SocketDevState::Connecting;
+
+        let now = Instant::now();
+
+        // Retry connecting for 10 secs
+        loop {
+            match UnixStream::connect(socket_path) {
+                Ok(s) => {
+                    let fd = s.as_raw_fd();
+                    self.ctrl_fd = fd;
+                    self.stream = Some(s);
+                    self.state = SocketDevState::Connected;
+                    debug!("Connected to tpm socket path : {:?}", socket_path);
+                    return Ok(());
+                }
+                Err(_e) => {}
+            };
+            sleep(Duration::from_millis(1000));
```

Comment:
What thread does this run in? You can't do this from the vmm thread.

Reply:
I dropped the retry logic here. With multiple test confirmed, the retry logic isn't really required.